### DISCUSSION TITLE
How to import a user with an argon2id-hashed password

### DISCUSSION THREAD

**Discussion by ashep** [zitadel/zitadel#8195] [Q&A] [answered]
Hello there!

I have a self-hosted Zitadel v 2.52, and I am trying to import a user with a password hashed using `argon2id` algorithm. I get the following error:

```
Password hash encoding is not supported. Check out https://zitadel.com/docs/concepts/architecture/secrets#hashed-secrets (USER-JDk4t)
```

According to [this doc](https://zitadel.com/docs/concepts/architecture/secrets#hashed-secrets), `argon2id`-hashed password are supported by Zitadel, while they're disabled on Cloud.

Can you please point me out how to perform the import of such users. Thank you.

UPD: I use [this](https://zitadel.com/docs/apis/resources/user_service/user-service-add-human-user) method.

UPD2: example of the hash I use: `$argon2id$v=19$m=65536,t=4,p=1$RVloRHFtbktyTFZOSVIyMA$7WAudQ`

**Comment by ashep** [comment]
It seems I've found the answer [here](https://zitadel.com/docs/self-hosting/manage/configure).

```
    # Verifiers enable the possibility of verifying
    # passwords that are previously hashed using another
    # algorithm then the Hasher.
    # This can be used when migrating from one algorithm to another,
    # or when importing users with hashed passwords.
    # There is no need to enable a Verifier of the same algorithm
    # as the Hasher.
    #
    # The format of the encoded hash strings must comply
    # with https://github.com/P-H-C/phc-string-format/blob/master/phc-sf-spec.md
    # https://passlib.readthedocs.io/en/stable/modular_crypt_format.html
    #
    # Supported verifiers: (uncomment to enable)
    Verifiers: # ZITADEL_SYSTEMDEFAULTS_PASSWORDHASHER_VERIFIERS
    #   - "argon2"   # verifier for both argon2i and argon2id.
    #   - "bcrypt"
    #   - "md5"      # md5Crypt with salt and password shuffling.
    #   - "md5plain" # md5 digest of a password without salt
    #   - "scrypt"
    #   - "pbkdf2"   # verifier for all pbkdf2 hash modes.
```

But now the problem is how to pass this option using an env variable. When I try to specify:

```
ZITADEL_SYSTEMDEFAULTS_PASSWORDHASHER_VERIFIERS='["argon2", "bcrypt"]'
```

I get the error:

```
password hasher: ID=CRYPT-sahW9 Message=password hash config invalid Parent=(invalid verifier \"[\\\"argon\\\"\")
```

What is the correct syntax to specify this?

**Reply by fforootd** [reply]
maybe @eliobischof has at hand how that is being read 😁

**Reply by ashep** [reply, marked as answer]
@fforootd I've figured out how to do this. Using comma:

```
ZITADEL_SYSTEMDEFAULTS_PASSWORDHASHER_VERIFIERS='argon2,bcrypt'
```

**Reply by fforootd** [reply]
Perfect, thanks for sharing this.
